ssr-express/lib/createAmpSsrMiddleware.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/**
+ * The transform middleware replaces the `res.write` method so that, instead of sending
+ * the content to the network, it is accumulated in a buffer. `res.end` is also replaced
+ * so that, when it is invoked, the buffered response is transformed with AMP-SSR and sent
+ * to the network.
+ */
+
+const mime = require('mime-types');
+const AmpSsrUrlScheme = require('./AmpSsrUrlScheme.js');
+
+const DEFAULT_URL_SCHEME = new AmpSsrUrlScheme('amp');
+
+/**
+ * Creates a new amp-server-side-rendering middleware, using the specified
+ * ampSSR and options.
+ *
+ * @param {string} ampSsr the ampSsr instance to be used by this transformer.
+ * @param {Object} options an optional object containing custom configurations for
+ * the middleware.
+ * @param {AmpSsrUrlScheme} options.ampSsrUrlScheme The scheme to be used when checking
+ * for AMP pages, rewriting to Canonical and generating amphtml links.
+ */
+const createAmpSsrMiddleware = (ampSsr, options) => {
```

Comment:
How about creating a builder instead that provides a good default:

```
class AmpMiddleware {

  constructor() {
     this.ampSSR_ = ampSSR;
     this.options_ = {};
  }

  withAmpSsr(ampSsr) {
    this.ampSSR_ = ampSSR;
  }

  withOptions(options) {
    this.options_ = options;
  }

  create() {
    return (req, res, next) => {
       ...
    }
   }
}

module.exports = new AmpMiddleware();
```

You can then use it like:

```
const AmpMiddleware = require('amp-toolbox-ssr-express');

AmpMiddleware.create()

AmpMiddleware.withAmpSSR(...).create()
```

Reply:
We'll will also need an option for using versioned URLs.

Reply:
I like the builder, but most express middlewares (static, compression, bodyparser, etc) pass the configuration as an option when instantiating. I think it's worth keeping the API familiar to what other middlewares do.